crates/sandboxer/main.rs

```diff
@@ -0,0 +1,152 @@
+        match open(&path, OFlag::O_PATH | OFlag::O_CLOEXEC, Mode::empty()) {
+            Err(e) => {
+                eprintln!("Failed to open \"{}\": {}\n", path.to_string_lossy(), e);
+                return Err(());
+            }
+            Ok(parent_fd) => match fstat(parent_fd) {
+                Ok(stat) => {
+                    if (stat.st_mode & SFlag::S_IFMT.bits()) != SFlag::S_IFDIR.bits() {
+                        actual_access.insert(ACCESS_FILE);
+                    }
+
+                    ruleset = match ruleset.add_path_beneath_rule(parent_fd, actual_access) {
+                        Ok(r) => r,
+                        Err(e) => {
+                            eprintln!(
+                                "Failed to update ruleset with \"{}\": {}\n",
+                                path.to_string_lossy(),
+                                e
+                            );
+                            return Err(());
+                        }
+                    };
+                }
+                Err(e) => {
+                    eprintln!("Failed to stat \"{}\": {}\n", path.to_string_lossy(), e);
+                    return Err(());
+                }
+            },
+        }
+    }
+
+    Ok(ruleset)
+}
+
+fn main() {
+    let args: Vec<_> = env::args_os().collect();
+
+    let program_name = args.get(0).map(|s| s.to_string_lossy()).unwrap_or_default();
+
+    if args.len() < 2 {
+        eprintln!(
+            "usage: {}=\"...\" {}=\"...\" {} <cmd> [args]...\n",
+            ENV_FS_RO_NAME, ENV_FS_RW_NAME, program_name
+        );
+        eprintln!("Launch a command in a restricted environment.\n");
+        eprintln!("Environment variables containing paths, each separated by a colon:");
+        eprintln!(
+            "* {}: list of paths allowed to be used in a read-only way.",
+            ENV_FS_RO_NAME
+        );
+        eprintln!(
+            "* {}: list of paths allowed to be used in a read-write way.",
+            ENV_FS_RW_NAME
+        );
+        eprintln!(
+            "\nexample:\n\
+				     {}=\"/bin:/lib:/usr:/proc:/etc:/dev/urandom\" \
+				     {}=\"/dev/null:/dev/full:/dev/zero:/dev/pts:/tmp\" \
+				     {} bash -i",
+            ENV_FS_RO_NAME, ENV_FS_RW_NAME, program_name
+        );
+        return;
+    }
+
+    let fs_ro = env::var_os(ENV_FS_RO_NAME)
+        .expect(&format!("Missing environment variable {}", ENV_FS_RO_NAME));
+    let fs_rw = env::var_os(ENV_FS_RW_NAME)
+        .expect(&format!("Missing environment variable {}", ENV_FS_RW_NAME));
+
+    let cmd_name = args
+        .get(1)
+        .map(|s| s.to_string_lossy())
+        .unwrap_or(Cow::Borrowed("bash -i"));
+
+    let ruleset = match RulesetAttr::new().create() {
+        Err(_) => {
+            eprintln!("Failed to create a ruleset");
+            return;
+        }
+        Ok(ruleset) => ruleset,
+    };
+
+    let ruleset = match populate_ruleset(ruleset, fs_ro, ACCESS_FS_ROUGHLY_READ) {
```

Comment:
It would be cleaner to use `?` and then convert `main()` to return a `Result`: https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/question-mark-in-main-and-tests.html